### DISCUSSION TITLE
ANDROID: extremely long automation actions

### DISCUSSION THREAD

**Discussion by mastafungus** [appium/appium#20614] [General]
We are testing a native application built with MAUI.

The tests are executed on 3 platforms:
ANDROID: Appium + UiAutomator2
iOS: Appium + XCUITest
WINDOWS: Appium + WinAppDriver

We have been encountering an issue on ANDROID for several months.

Specifically, there are 3 pages in the application where automation actions, such as clicking or text input, take an excessive amount of time without causing the test to fail. It can take up to 1 to 2 minutes for a simple click.
Only 3 pages are affected, and this issue occurs exclusively on ANDROID.

Testing is performed on physical devices. ANDROID and WINDOWS are local, while iOS is remote.
The application is relatively complex, but the elements we wish to interact with on the three pages in question are immediately visible, and there are no ongoing loading processes.

I should add that we have been working on this application for over a year now, and the issue has only presented itself in the last few months.
Additionally, manual actions do not exhibit any sluggishness; the slowdown occurs only during the automation of these actions.

Has anyone encountered this problem before?

**Comment by KazuCocoa** [comment]
https://github.com/appium/appium-uiautomator2-driver?tab=readme-ov-file#poor-elements-interaction-performance could help, although if it helped, maybe the app's main thread usage could be improved.

**Comment by mastafungus** [comment]
Thank you very much KazuCocoa!

Applying the waitForIdleTimeout parameter with a timeout reduced to ANDROID's capabilities has clearly improved the speed of test execution and, most importantly, has significantly reduced the problematic action times I was encountering on 3 troublesome pages. Now, actions take between 1 and 5 seconds on these pages, whereas they used to take between 1 and 3 minutes before!
